modules/core/src/main/java/org/apache/synapse/util/ConcurrencyThrottlingUtils.java

```diff
@@ -0,0 +1,69 @@
+/*
+ *  Copyright (c) 2018 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied. See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ *
+ */
+package org.apache.synapse.util;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.SynapseConstants;
+import org.apache.synapse.commons.throttle.core.ConcurrentAccessController;
+import org.apache.synapse.commons.throttle.core.ConcurrentAccessReplicator;
+
+/**
+ * Utility class with related Synapse concurrency throttling.
+ */
+public class ConcurrencyThrottlingUtils {
+
+    private static final Log log = LogFactory.getLog(ConcurrencyThrottlingUtils.class);
+
+    /**
+     * Decrement the internal counter for concurrency throttling, in case of normal response retrieval,
+     * delayed response at synapse timeout, exceptional cases while mediation end up trigger fault.
+     * In clustered environment, replicate the decremented value to other members.
+     *
+     * @param synCtx Synapse Message Context of which mediation occurs.
+     */
+    public static void decrementConcurrencyThrottleAccessController(MessageContext synCtx) {
+
+        Boolean isConcurrencyThrottleEnabled = (Boolean) synCtx
+                .getProperty(SynapseConstants.SYNAPSE_CONCURRENCY_THROTTLE);
+
+        if (isConcurrencyThrottleEnabled != null && isConcurrencyThrottleEnabled) {
+            ConcurrentAccessController concurrentAccessController = (ConcurrentAccessController) synCtx
+                    .getProperty(SynapseConstants.SYNAPSE_CONCURRENT_ACCESS_CONTROLLER);
+            int available = concurrentAccessController.incrementAndGet();
+            int concurrentLimit = concurrentAccessController.getLimit();
```

Comment:
If available and concurrentLimit just used only for debug log, please move them inside if (log.isDebugEnabled()) block, current usage is waste of processing and context switching

Reply:
+1 Done